File: app/csharp/AccountingSystem.Api/Infrastructure/Web/Controllers/AuditLogController.cs
using AccountingSystem.Application.Ports.In;
using AccountingSystem.Domain.Models.Audit;
using AccountingSystem.Infrastructure.Web.Dtos;
using Microsoft.AspNetCore.Mvc;

namespace AccountingSystem.Infrastructure.Web.Controllers;

/// <summary>
/// 監査ログ REST API コントローラー
/// </summary>
[ApiController]
[Route("api/v1/audit-logs")]
[Tags("監査ログAPI")]
public class AuditLogController : ControllerBase
{
    private readonly IAuditLogService _auditLogService;

    public AuditLogController(IAuditLogService auditLogService)
    {
        _auditLogService = auditLogService;
    }

    /// <summary>
    /// エンティティの変更履歴を取得する
    /// </summary>
    /// <param name="entityType">エンティティ種別（Account, Journal等）</param>
    /// <param name="entityId">エンティティID</param>
    /// <returns>監査ログ一覧</returns>
    [HttpGet("entity/{entityType}/{entityId}")]
    [ProducesResponseType(typeof(List<AuditLogResponse>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetEntityHistory(
        string entityType,
        string entityId)
    {
        var logs = await _auditLogService.GetEntityHistoryAsync(entityType, entityId);
        return Ok(logs.Select(AuditLogResponse.From).ToList());
    }

    /// <summary>
    /// ユーザーの操作履歴を取得する
    /// </summary>
    /// <param name="userId">ユーザーID</param>
    /// <param name="startDate">開始日時</param>
    /// <param name="endDate">終了日時</param>
    /// <returns>監査ログ一覧</returns>
    [HttpGet("user/{userId}")]
    [ProducesResponseType(typeof(List<AuditLogResponse>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetUserActivity(
        string userId,
        [FromQuery] DateTime startDate,
        [FromQuery] DateTime endDate)
    {
        var logs = await _auditLogService.GetUserActivityAsync(userId, startDate, endDate);
        return Ok(logs.Select(AuditLogResponse.From).ToList());
    }

    /// <summary>
    /// 期間別の監査ログを取得する
    /// </summary>
    /// <param name="startDate">開始日時</param>
    /// <param name="endDate">終了日時</param>
    /// <param name="limit">取得件数上限（デフォルト: 100）</param>
    /// <returns>監査ログ一覧</returns>
    [HttpGet("period")]
    [ProducesResponseType(typeof(List<AuditLogResponse>), StatusCodes.Status200OK)]
    public async Task<IActionResult> GetAuditLogsByPeriod(
        [FromQuery] DateTime startDate,
        [FromQuery] DateTime endDate,
        [FromQuery] int limit = 100)
    {
        var logs = await _auditLogService.GetAuditLogsForPeriodAsync(startDate, endDate, limit);
        return Ok(logs.Select(AuditLogResponse.From).ToList());
    }

    /// <summary>
    /// アクション別の監査ログを取得する
    /// </summary>
    /// <param name="action">アクション種別（CREATE, UPDATE, DELETE）</param>
    /// <param name="startDate">開始日時</param>
    /// <param name="endDate">終了日時</param>
    /// <param name="limit">取得件数上限（デフォルト: 100）</param>
    /// <returns>監査ログ一覧</returns>
    [HttpGet("action/{action}")]
    [ProducesResponseType(typeof(List<AuditLogResponse>), StatusCodes.Status200OK)]
    [ProducesResponseType(typeof(ErrorResponse), StatusCodes.Status400BadRequest)]
    public async Task<IActionResult> GetAuditLogsByAction(
        string action,
        [FromQuery] DateTime startDate,
        [FromQuery] DateTime endDate,
        [FromQuery] int limit = 100)
    {
        if (!Enum.TryParse<AuditAction>(action, true, out var auditAction))
        {
            return BadRequest(new ErrorResponse
            {
                Error = "INVALID_ACTION",
                Message = $"無効なアクション種別です: {action}。有効な値: CREATE, UPDATE, DELETE"
            });
        }

        var logs = await _auditLogService.GetAuditLogsByActionAsync(auditAction, startDate, endDate, limit);
        return Ok(logs.Select(AuditLogResponse.From).ToList());
    }
}

File: app/fsharp/AccountingSystem.Infrastructure/Web/Controllers/AuditLogController.fs
namespace AccountingSystem.Infrastructure.Web.Controllers

open System
open System.Threading.Tasks
open Microsoft.AspNetCore.Http
open Microsoft.AspNetCore.Mvc
open AccountingSystem.Application.Port.In
open AccountingSystem.Infrastructure.Web.Dtos

/// <summary>
/// 監査ログ REST API コントローラー（Input Adapter）
/// </summary>
[<ApiController>]
[<Route("api/v1/audit-logs")>]
[<Tags("監査ログAPI")>]
type AuditLogController(auditLogService: IAuditLogUseCase) =
    inherit ControllerBase()

    /// <summary>
    /// エンティティ別の監査ログを取得
    /// </summary>
    [<HttpGet("entity/{entityType}/{entityId}")>]
    [<ProducesResponseType(typeof<AuditLogResponse list>, StatusCodes.Status200OK)>]
    member _.GetByEntity(entityType: string, entityId: string) : Task<IActionResult> =
        task {
            let! logs = auditLogService.GetByEntityAsync(entityType, entityId)
            let response = logs |> List.map AuditLogResponse.fromDomain
            return OkObjectResult(response) :> IActionResult
        }

    /// <summary>
    /// ユーザー別の監査ログを取得
    /// </summary>
    [<HttpGet("user/{userId}")>]
    [<ProducesResponseType(typeof<AuditLogResponse list>, StatusCodes.Status200OK)>]
    member _.GetByUser(
        userId: string,
        [<FromQuery>] startDate: DateTime,
        [<FromQuery>] endDate: DateTime) : Task<IActionResult> =
        task {
            let! logs = auditLogService.GetByUserAsync(userId, startDate, endDate)
            let response = logs |> List.map AuditLogResponse.fromDomain
            return OkObjectResult(response) :> IActionResult
        }

    /// <summary>
    /// 期間内のすべての監査ログを取得
    /// </summary>
    [<HttpGet>]
    [<ProducesResponseType(typeof<AuditLogResponse list>, StatusCodes.Status200OK)>]
    member _.GetByDateRange(
        [<FromQuery>] startDate: DateTime,
        [<FromQuery>] endDate: DateTime) : Task<IActionResult> =
        task {
            let! logs = auditLogService.GetByDateRangeAsync(startDate, endDate)
            let response = logs |> List.map AuditLogResponse.fromDomain
            return OkObjectResult(response) :> IActionResult
        }

    /// <summary>
    /// アクション種別で監査ログを検索
    /// </summary>
    [<HttpGet("action/{actionType}")>]
    [<ProducesResponseType(typeof<AuditLogResponse list>, StatusCodes.Status200OK)>]
    member _.GetByAction(
        actionType: string,
        [<FromQuery>] startDate: DateTime,
        [<FromQuery>] endDate: DateTime) : Task<IActionResult> =
        task {
            let! logs = auditLogService.GetByActionAsync(actionType, startDate, endDate)
            let response = logs |> List.map AuditLogResponse.fromDomain
            return OkObjectResult(response) :> IActionResult
        }

    /// <summary>
    /// 監査ログを記録（通常はシステム内部から呼び出される）
    /// </summary>
    [<HttpPost>]
    [<ProducesResponseType(typeof<AuditLogResponse>, StatusCodes.Status201Created)>]
    [<ProducesResponseType(typeof<ErrorResponse>, StatusCodes.Status400BadRequest)>]
    member this.RecordAuditLog([<FromBody>] request: CreateAuditLogRequest) : Task<IActionResult> =
        task {
            match CreateAuditLogRequest.toDomain request with
            | Some auditLog ->
                let! recorded = auditLogService.RecordAsync(auditLog)
                return
                    this.CreatedAtAction(
                        nameof this.GetByEntity,
                        {| entityType = recorded.EntityType; entityId = recorded.EntityId |},
                        AuditLogResponse.fromDomain recorded) :> IActionResult
            | None ->
                return BadRequestObjectResult({| Error = "無効なアクション種別です" |}) :> IActionResult
        }
